### DISCUSSION TITLE
Deciding whether to return a response

### DISCUSSION THREAD

**Discussion by bcb** [bcb/jsonrpcserver#163] [General]
There's one more concern with JSON-RPC -- you shouldn't always return a response.

There are two reasons a response should **not** be sent:

- The request was a notification (no id was included).
- The request was a batch, and all requests in the batch were notifications (no id was included in any).

The question is how the developer API to handle this should look.

I'm thinking of a `should_respond` function which makes the decision for you based on the return value from dispatch, something like this (note synchronous protocols like http must return something)

```python
from jsonrpcserver.response import should_respond

@app.route("/", methods=["POST"])
def index():
    response = dispatch(request.get_data().decode())
    return to_json(response) if should_respond(response) else ""
```

For async protocols something like:

```python
response = async dispatch(request)
if should_respond(response):
    send(to_json(response))
```

Does this make sense? Would appreciate any thoughts.

**Comment by bcb** [comment]
Also, to be very strict, `to_json` would raise an exception if the request was a notification or batch of notifications. This way you _must_ use should_respond before using to_json.

**Comment by st31ny** [comment]
If I understand the JSON-RPC specification correctly, the distinction between requests and notifications lies within the client only. So, while it might make sense in some cases in the server implementation to know whether a response is required (e.g., to avoid computationally intense calculations), I'd rather want the server implementation to handle that part of the communication alone.
TL;DR: I think server method implementations should not need to worry about the request/notification distinction. This should be handled in the server.

**Reply by bcb** [reply]
Yeah that’s right, the method doesn’t worry about this, but the server implementation needs to know whether or not it should respond to the client. If the request was a notification (or a list of notifications), it shouldn’t respond. Having a should_respond function is the only way i can think to do this.

**Reply by st31ny** [reply]
Why not handle in `dispatch()`?

EDIT: I think I didn't get this in the beginning. So the `should_respond` is required in the code calling dispatch, not in the server implementation. Then I agree, something like this is probably needed. Only other option I do see is to use `None` to signal that no response is required (or a special response type like `NoResponse`).

**Reply by bcb** [reply]
Because dispatch doesn’t send a response back to the client, dispatch only gives you a `Union[Response, List[Response]]`

It’s up to the implementation to send a response. (different protocols send a responses in different ways, impossible for jsonrpcserver to do this)

**Reply by bcb** [reply]
The spec says “The Server MUST NOT reply to a Notification”

**Reply by st31ny** [reply]
Yeah, I see, so what about an empty list?

**Reply by bcb** [reply]
Respond with an empty list? That would break the spec which says we mustn’t respond at all. Even responding with an empty string is wrong but in http you have to respond with something.

**Reply by st31ny** [reply]
No, I meant as response from `dispatch()` such that the caller knows that there is nothing to respond. But I don't really like that idea either. I guess it's better to have a function to ask.

**Reply by bcb** [reply]
dispatch gives you a Union[Response, List[Response]]

Which you pass to to_json to convert it into a string to respond with.

I could make to_json give an empty string for notifications but I feel that’s incorrect, and it doesn’t help e.g. websockets that needs to know whether to send a response back or do nothing at all.

**Reply by bcb** [reply]
It’s not very nice to have

dispatch -> should_respond? -> to_json

I’d rather remove should respond but can’t think of another way.

**Reply by st31ny** [reply]
Yes, I agree. What about a `None` result? So `to_json()` returns `None`. Not very nice either but saves the `if`.

**Reply by bcb** [reply]
But then wouldn’t you still have to check the to_json result

**Reply by st31ny** [reply]
True. At least one check is required I guess since somebody needs to decide whether to send something or not at the end of the day.

**Reply by bcb** [reply]
Yeah, there’s no avoiding it. 

It feels clunky but that’s the spec.

**Comment by hholst80** [comment]
I think one can avoid introducing a API method for this. Consider this:

```python
@app.route("/", methods=["POST"])
def handler():
    response = dispatch(...)
    if response is None:
        return Response(status_code=HTTP_204_NO_CONTENT)
    else:
        return to_json(response)   # it must have been something we should reply to, error or not, and it has an Id.
```

The empty string `""` is a proper response but if the `content-type` is `application/json` it could end up being interpreted as a completely empty document, which probably is not allowed by the spec (how would it be interpreted - how do you return Nothing in JSON?).

Ideas found here: https://github.com/tiangolo/fastapi/issues/717

If `to_json` throws an exception it's a bad thing, but it is not the responsibility of this library to solve it. ;-)

**Reply by bcb** [reply]
Yes true, we could use None and then won't need should_respond.

But what do you think `to_json` should do if `response` is `None`?

Because a lot of people will not check the result of `dispatch`.

**Reply by bcb** [reply]
To me to_json should raise an exception if it’s passed `None`.

**Reply by bcb** [reply]
Well, json.dumps returns "null" when passed None, so that will do.

**Reply by st31ny** [reply]
Which is technically correct since JSON knows the value `null` and Python maps its `None` value to JSON `null`.

**Reply by bcb** [reply]
Yep.

Ok, so in an asynchronous protocol you should check the result of dispatch, and only send a response if the result is not None. But in HTTP it's not really necessary. You can just say
```python
@app.route("/", methods=["POST"])
def index():
    return to_json(dispatch(request.get_data().decode()))
```

Which will return "null" to a notification.

**Reply by st31ny** [reply]
Looks good to me!

**Reply by bcb** [reply]
Maybe dispatch could just return the json, so there's no need for `to_json`.

In an async protocol like Websockets, you could check if the result is "null" and not respond if so.

I might make another topic about this.

**Reply by bcb** [reply]
Update: So we can't respond with "null" because although it's valid json, it's not valid json-rpc (the client will validate it against a json-rpc schema)

So it'll have to be an empty string (technically not valid json nor json-rpc, but we hope the client ignores this response)
